### PR TITLE
Add osm tag file parsing

### LINKED ISSUE #19: Investigate using osmfilter/osmium for generating inputs
As [discussed in #6](https://github.com/organicmaps/wikiparser/pull/6#discussion_r1253465466), 
Instead of waiting for the generator to reach the descriptions stage, we could process the OSM globe file directly to query the wikipedia and wikidata tags.

To separate them, we should be careful to:

- use the same criteria as the generator for filtering nodes
- process article titles in the same method as the generator

## Previous Work

I got a query with `osmfilter` working based the filters in [`ftypes_matcher.cpp`](https://github.com/organicmaps/organicmaps/blob/982c6aa92d7196a5690dcdc1564e427de7611806/indexer/ftypes_matcher.cpp#L473).

```
osmfilter planet.o5m --keep="( wikipedia= or wikidata= ) and ( amenity=grave_yard or amenity=fountain or amenity=place_of_worship or amenity=theatre or amenity=townhall or amenity=university or boundary=national_park or building=train_station or highway=pedestrian or historic=archaeological_site or historic=boundary_stone or historic=castle or historic=fort or historic=memorial or historic=monument or historic=ruins or historic=ship or historic=tomb or historic=wayside_cross or historic=wayside_shrine or landuse=cemetery or leisure=garden or leisure=nature_reserve or leisure=park or leisure=water_park or man_made=lighthouse or man_made=tower or natural=beach or natural=cave_entrance or natural=geyser or natural=glacier or natural=hot_spring or natural=peak or natural=volcano or place=square or tourism=artwork or tourism=museum or tourism=gallery or tourism=zoo or tourism=theme_park or waterway=waterfall or tourism=viewpoint or tourism=attraction )" \
| osmconvert - --csv-headline --csv="@oname @id wikipedia wikidata"
```

I ran it on the Yukon territory map and found that it output additional articles compared to the generator.

## Next Steps

- [x] Investigate earlier processing layers in map generator to improve the query.
- [x] Try to convert `osmfilter` command to `osmium` so `.pbf` files can be used directly.
- [x] Run query on whole planet file and compare with generator output and all wikipedia/wikidata tags in the planet file.
- [x] Update wikiparser to handle direct OSM inputs.

### PR DESCRIPTION
**Pull request by newsch**
Parse wikipedia and wikidata tags from a tsv file of OSM tags,
compatible with the `--csv` output of `osmconvert`.

Closes #19.

Notes from parsing all planet wikipedia/wikidata tags:
- 6 UTF-8 TSV parsing errors
- 585 tag parsing errors
    - 511 titles
        - no lang
        - `;`/`း` instead of `:`
        - qid instead of title
    - 74 qids
        - wikipedia title instead of qid
        - `Q123;Q124`
        - `(Q123)`
        - `Warfstermolen (Q1866088)`
        - yandex urls
        - amenity
        - coordinates

There are 50 wikipedia entries with url escaping, some are urls instead of titles and not handled correctly.

Remaining work:
- [x] add url checks in title parsing, match generator url handling (and handle mobile website urls)
- [x] ~~serialize parse errors to disk for changes~~ Add structured parse errors (see #25 for rest), log summary
- [x] Use `osmpbf` crate to parse planet file to fix `osmconvert` truncation problem
- [x] Merge #21 and update `run.sh` to use new method

### CODE REVIEW

**Comment by biodranik** [comment]
Is there an invalid UTF in OpenStreetMap.org? Any examples? Can it be fixed?

It would be great to fix errors in OpenStreetMap.org directly. 

On the other hand, it is important to log errors with OpenStreetMap.org data and report them to osmers or fix ourselves, while producing robust output.

**Comment by newsch** [comment]
I'm not sure yet if it's in the OSM db or something that got messed up by `osmium`/`osmconvert`, and I can't copy/paste them here, but I attached the relevant lines here:
[planet-wikipedia-utf8-errors.csv](https://github.com/organicmaps/wikiparser/files/12253829/planet-wikipedia-utf8-errors.csv)

```
CSV parse error: record 602681 (line 602682, field: 6, byte: 48946539): invalid utf-8: invalid UTF-8 in field 6 near byte index 25
CSV parse error: record 1326111 (line 1326112, field: 4, byte: 102037202): invalid utf-8: invalid UTF-8 in field 4 near byte index 25
CSV parse error: record 1610617 (line 1610618, field: 6, byte: 118698840): invalid utf-8: invalid UTF-8 in field 6 near byte index 25
CSV parse error: record 1778686 (line 1778687, field: 6, byte: 128580834): invalid utf-8: invalid UTF-8 in field 6 near byte index 25
CSV parse error: record 2279654 (line 2279713, field: 6, byte: 158513245): invalid utf-8: invalid UTF-8 in field 6 near byte index 25
CSV parse error: record 2723754 (line 2723813, field: 4, byte: 186764436): invalid utf-8: invalid UTF-8 in field 4 near byte index 254
```

I'll add an option to output the errors in a structured way that we can deal with. Some are straightforward fixes, others might require someone with local knowledge that we can leave notes on.

**Comment by biodranik** [comment]
Thanks! Could it be some local locale issue? What is your terminal locale/code page? Does it support non-US ones? Or it could be some osmium issue.

There is nothing wrong with the output: 
https://www.openstreetmap.org/api/0.6/node/1648348163
https://www.openstreetmap.org/node/1648348163

https://www.openstreetmap.org/api/0.6/way/53580264

https://www.openstreetmap.org/api/0.6/relation/7032757

**Comment by newsch** [comment]
My terminal is set to `en_US.UTF-8`, but it went straight from `osmium` -> `osmconvert` -> file.
I have an outdated planet file, I'll extract the full objects from there and see.

**Comment by newsch** [comment]
`osmium` outputs it fine, but `osmconvert` doesn't.
I think it's a truncation issue, they're all around 257 bytes long, the error is at the end, and [`osmconvert` has a fixed buffer of length 256 for keys/values](https://gitlab.com/osm-c-tools/osmctools/-/blob/f341f5f237737594c1b024338f0a2fc04fabdff3/src/osmconvert.c#L3177).

It looks like these aren't titles but notes? or copied text from the articles? [Wikipedia has a limit of 255 characters in titles.](https://ru.wikipedia.org/wiki/%D1%81.%20%D0%A5%D0%B0%D1%80%D0%B1%D0%B0%D0%BB%D0%B0%201-%D1%8F%20-%20%D1%81%D0%B5%D0%BB%D1%8C%D1%81%D0%BA%D0%B8%D0%B9%20%D0%BD%D0%B0%D1%81%D0%B5%D0%BB%D0%B5%D0%BD%D0%BD%D1%8B%D0%B9%20%D0%BF%D1%83%D0%BD%D0%BA%D1%82%20(%D1%81%D0%B5%D0%BB%D0%BE),%20%D0%B5%D0%B4%D0%B8%D0%BD%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D1%8B%D0%B9%20%D0%BD%D0%B0%D1%81%D0%B5%D0%BB%D0%B5%D0%BD%D0%BD%D1%8B%D0%B9%20%D0%BF%D1%83%D0%BD%D0%BA%D1%82%20%D0%B8%20%D0%B0%D0%B4%D0%BC%D0%B8%D0%BD%D0%B8%D1%81%D1%82%D1%80%D0%B0%D1%82%D0%B8%D0%B2%D0%BD%D1%8B%D0%B9%20%D1%86%D0%B5%D0%BD%D1%82%D1%80%20%D0%A1%D0%9F%20%C2%AB%D0%A5%D0%B0%D1%82%D1%8B%D0%BB%D1%8B%D0%BD%D1%81%D0%BA%D0%B8%D0%B9%20%D0%BD%D0%B0%D1%81%D0%BB%D0%B5%D0%B3%C2%BB%20%D0%A7%D1%83%D1%80%D0%B0%D0%BF%D1%87%D0%B8%D0%BD%D1%81%D0%BA%D0%BE%D0%B3%D0%BE%20%D1%83%D0%BB%D1%83%D1%81%D1%83%20(%D1%80%D0%B0%D0%B9%D0%BE%D0%BD%D0%B0)%20%D0%A0%D0%B5%D1%81%D0%BF%D1%83%D0%B1%D0%BB%D0%B8%D0%BA%D0%B8%20%D0%A1%D0%B0%D1%85%D0%B0%20(%D0%AF%D0%BA%D1%83%D1%82%D0%B8%D1%8F),%20%D1%80%D0%B0%D1%81%D0%BF%D0%BE%D0%BB%D0%BE%D0%B6%D0%B5%D0%BD%D0%BD%D1%8B%D0%B9%20%D0%B2%207%20%D0%BA%D0%B8%D0%BB%D0%BE%D0%BC%D0%B5%D1%82%D1%80%D0%B0%D1%85%20%D0%BE%D1%82%20%D0%A7%D1%83%D1%80%D0%B0%D0%BF%D1%87%D0%B8,%20%D0%B0%D0%B4%D0%BC%D0%B8%D0%BD%D0%B8%D1%81%D1%82%D1%80%D0%B0%D1%82%D0%B8%D0%B2%D0%BD%D0%BE%D0%B3%D0%BE%20%D1%86%D0%B5%D0%BD%D1%82%D1%80%D0%B0)

**Comment by biodranik** [comment]
Good catch! You can patch osmconvert to increase the size, or try native rust approach.

**Comment by biodranik** [comment]
The link above says 255 urf-8 bytes in title, not 255 characters. 

You may try to osmupdate your planet (don't forget to drop authors and history).

**Comment by newsch** [comment]
Sorry, I misspoke - yes, 255 bytes of utf-8. I'll try the rust parser you used.

**Review by biodranik — COMMENTED** [review]
Such a big PR! It would be easier to do it in parts.

How does simplification output look now?

Can we run and test it?

**Comment by newsch** [comment]
> Such a big PR! It would be easier to do it in parts.

I will try to break it up into better commits and let you know - some of it is refactoring that isn't helpful to see together.

> How does simplification output look now?

This doesn't touch simplification, I am doing that next. I'll add my updates to #4 and open a PR with the changes.

> Can we run and test it?

Sure! I've been testing it as I go, and `run.sh` is updated to extract the tags from a pbf file. The new format is documented in the script and the README, the TL;DR is that now you pass a pbf file as well:
```
./run.sh /tmp/output/ ~/Downloads/planet-latest.osm.pbf ~/Downloads/enwiki-NS0-20230401-ENTERPRISE-HTML.json.tar.gz ...
```

File: benches/id_parsing.rs [hunk 1]
```diff
@@ -7,19 +7,16 @@ extern crate test;
 #[bench]
 fn parse_wikipedia(b: &mut test::Bencher) {
     b.iter(|| {
-        let title = om_wikiparser::wm::WikipediaTitleNorm::from_url(
-            "https://en.wikipedia.org/wiki/Article_Title",
-        )
-        .unwrap();
+        let title =
+            om_wikiparser::wm::Title::from_url("https://en.wikipedia.org/wiki/Article_Title")
```
[the remark this answers is not in the document]
Comment (newsch) [reply benches/id_parsing.rs:11]: I'm not sure, renaming it to the shorter `Title` must have altered `rustfmt`'s heuristics.

**Comment by newsch** [comment]
I think it will be easiest to review the remaining commits individually. They're reasonably separated and each has a meaningful commit message.

These contain meaningful changes:
- [Add new option to parse osm tag file](https://github.com/organicmaps/wikiparser/pull/23/commits/a2c113a8850ced2c115fe2065ad169956ece49fa) (+114/-11)
- [Extract tags in parallel in rust](https://github.com/organicmaps/wikiparser/pull/23/commits/3d48c39793cc39129d37d0e07851c2488160f3c4) (+559/-97)
- [Improve url handling](https://github.com/organicmaps/wikiparser/pull/23/commits/2532d1365e5d50e77e533963df27f6465f6618cc) (+26/-2)
- [Structure parse errors and only log warning if above threshold](https://github.com/organicmaps/wikiparser/pull/23/commits/b250dd4b1390bf23c673923f4e9a65f8c6d3a32e) (+155/-40)

These move things around and don't change meaningful functionality:
- [Refactor into subcommands](https://github.com/organicmaps/wikiparser/pull/23/commits/0ac935c175a161a1e60ac7ebe917d858721df308) (+305/-277)
- [Refactor and rename title/qid wrappers](https://github.com/organicmaps/wikiparser/pull/23/commits/29cdbe2301dce8a615505ba4442208e50127f33d) (+211/-207)

File: src/wm/title.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        let url = Url::parse(url.trim())?;
+
+        let (subdomain, host) = url
+            .host_str()
+            .ok_or_else(|| anyhow!("Expected host"))?
+            .split_once('.')
+            .ok_or_else(|| anyhow!("Expected subdomain"))?;
+        let host = host.strip_prefix("m.").unwrap_or(host);
+        if host != "wikipedia.org" {
+            bail!("Expected wikipedia.org for domain")
```
Comment (biodranik) [inline src/wm/title.rs]: Does it make sense to print wrong hosts in a log to fix/support them?

File: src/wm/title.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        }
+        let lang = subdomain;
+
+        let path = url.path();
+
+        let (root, title) = path
+            .strip_prefix('/')
+            .unwrap_or(path)
+            .split_once('/')
+            .ok_or_else(|| anyhow!("Expected at least two segments in path"))?;
```
Comment (biodranik) [inline src/wm/title.rs]: ditto

File: src/get_articles.rs [hunk 4]
```diff
@@ -60,9 +60,34 @@ pub fn run(args: Args) -> anyhow::Result<()> {
...
+            path,
+            &mut wikidata_qids,
+            &mut wikipedia_titles,
+            Some(&mut line_errors),
+        )?;
+
+        if !line_errors.is_empty() {
+            let error_count = line_errors.len();
+            let new_items = wikidata_qids.len() + wikipedia_titles.len() - original_items;
+            let expected_threshold = 0.02;
```
Comment (biodranik) [inline src/get_articles.rs:78]: What is the benefit of hiding errors under a threshold? Isn't it beneficial to see all errors and be able to estimate/compare the quality of the dump, and to easily grep/find what is most important, or feed the whole log to contributors for fixes?

**Review by biodranik — APPROVED** [review]
Thanks for clear commits and their descriptions!

File: src/wm/title.rs (hunk 2 above)
Comment (newsch) [reply src/wm/title.rs]: They are caught at a higher level and logged/saved with the full string

File: src/get_articles.rs (hunk 4 above)
Comment (newsch) [reply src/get_articles.rs:78]: The threshold only determines if the message is `info` vs `error` level.
When you use the `run.sh` script with multiple languages it prints a copy of the hundreds of errors for each language.
I think writing the parse errors to a file separately will be easier to read and deal with.

I'm open to other ideas.

**Pull request merged**